File: models/article.go
package models // Package models 模型

import (
	"time"

	"gorm.io/gorm"
)

type Category struct {
	gorm.Model
	Name string `json:"name"` // 分类名称
}
type Tag struct {
	gorm.Model
	Name     string    `json:"name" gorm:"unique"` // 标签名称
	Articles []Article `gorm:"many2many:article_tags;"`
}

type Article struct {
	gorm.Model
	Title    string    `json:"title" binding:"required"`   //标题
	Content  string    `json:"content" binding:"required"` //内容
	Preview  string    `json:"preview" binding:"required"` //预览
	Likes    int       `json:"likes" gorm:"default:0"`     //点赞数，默认为0
	Views    int       `json:"views" gorm:"default:0"`     //浏览数，默认为0
	User     *User     `json:"user" gorm:"foreignKey:UserID"`
	UserID   uint      `json:"user_id" binding:"required"`
	CoverImg string    `json:"cover_img"`                                                         //【新增】封面图的 URL
	Comments []Comment `json:"comments" gorm:"foreignKey:ArticleID;constraint:OnDelete:CASCADE;"` //评论

	CategoryID uint      `json:"category_id"`                           // 记录分类的 ID
	Category   *Category `json:"category" gorm:"foreignKey:CategoryID"` //类别

	Tags []Tag `json:"tags" gorm:"many2many:article_tags;"`

	FavoredBy []User `gorm:"many2many:user_article_favor"` //被哪些用户收藏
}

type ArticleTags struct {
	TagID     uint `gorm:"primaryKey"`
	ArticleID uint `gorm:"primaryKey"`
	CreatedAt time.Time
}

type Comment struct {
	gorm.Model
	ArticleID uint   `json:"article_id" binding:"required"` // 外键，关联到 Article.ID
	UserID    uint   `json:"user_id" binding:"required"`    // 评论者 ID
	Content   string `json:"content" binding:"required"`    // 评论具体内容
}

File: frontend/src/types/article.ts
/** 文章列表项 —— 严格匹配后端 ArticleListResponse 结构 */
export interface ArticleListItem {
    id: number
    title: string
    preview: string
    likes: number
    views: number
    user_id: number
    username: string      // 作者用户名
    category_name: string
    tags: string[]       // 后端列表接口返回字符串数组，如 ["cs", "go"]
    cover_img: string
    created_at: string
}

/** 文章详情 —— 匹配后端 models.Article 结构 */
export interface ArticleDetail {
    id: number
    title: string
    content: string
    preview: string
    likes: number
    views: number
    user_id: number
    cover_img: string
    category_id: number | null
    category: { id: number; name: string; slug?: string } | null
    tags: Array<{ id: number; name: string }>
    comments: Comment[]
    user: Author             // 后端 json tag 是 "user"
    favorited?: boolean      // 后端可能不返回此字段
    favorites_count?: number // 后端可能不返回此字段
    created_at: string
    updated_at: string
}

/** 文章创建请求 —— 一级扁平结构，tag_ids 为纯数字数组 */
export interface ArticleCreateRequest {
    title: string
    content: string
    preview: string
    category_id: number | null
    tag_ids: number[]      // 🔥 锁死为纯数字数组
    cover_img: string
}

/** 文章更新请求 */
export interface ArticleUpdateRequest {
    title: string
    content: string
    preview: string
    category_id: number
    tag_ids: number[]      // 🔥 纯数字数组
    cover_img: string
}

/** 后端文章列表响应结构 */
export interface ArticlesResponse {
    data: ArticleListItem[]
    page: number
    page_size: number
    total: number
    total_pages: number
}

/** 后端文章详情响应结构 */
export interface ArticleDetailResponse {
    article: ArticleDetail
    comments: Comment[]
    likes: string    // 后端以字符串返回点赞数
}

/** 后端创建/更新文章响应结构 */
export interface ArticleMutationResponse {
    id: number
    title: string
    content: string
    preview: string
    category_id: number | null
    cover_img: string
    user_id: number
    created_at: string
    updated_at: string
}

export interface Author {
    id: number
    username: string
    bio: string
    image: string
    following: boolean
    created_at: string
    updated_at: string
}

export interface Comment {
    id: number
    article_id: number
    user_id: number
    content: string
    author?: Author
    created_at: string
}

/** 后端评论列表响应结构 */
export interface CommentsResponse {
    comments: Comment[]
}

/** 创建评论请求 */
export interface CommentCreateRequest {
    content: string
}

/** 创建评论响应（返回创建的评论对象） */
export type CommentResponse = Comment

/** 文章详情响应（兼容别名） */
export type ArticleResponse = ArticleDetailResponse

/** 文章列表响应（兼容别名） */
export type ArticleListResponse = ArticlesResponse
